File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/overlay/overlay/NCircleOverlay.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.overlay.overlay

import androidx.annotation.ColorInt
import com.naver.maps.geometry.LatLng
import com.naver.maps.map.overlay.CircleOverlay
import com.naver.maps.map.overlay.Overlay
import dev.note11.flutter_naver_map.flutter_naver_map.converter.AddableOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asDouble
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asInt
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asLatLng
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.toMap
import dev.note11.flutter_naver_map.flutter_naver_map.model.enum.NOverlayType
import dev.note11.flutter_naver_map.flutter_naver_map.model.overlay.NOverlayInfo
import dev.note11.flutter_naver_map.flutter_naver_map.util.DisplayUtil

internal data class NCircleOverlay(
    override val info: NOverlayInfo,
    val center: LatLng,
    val radius: Double, // meter
    @ColorInt val color: Int,
    @ColorInt val outlineColor: Int,
    val outlineWidthDp: Double,
) : AddableOverlay<CircleOverlay> {

    override fun createMapOverlay(): CircleOverlay = CircleOverlay().also { c ->
        c.center = center
        c.radius = radius
        c.color = color
        c.outlineColor = outlineColor
        c.outlineWidth = DisplayUtil.dpToPx(outlineWidthDp)
    }

    override fun toMap(): Map<String, Any?> = mapOf(
        infoName to info.toMap(),
        centerName to center.toMap(),
        radiusName to radius,
        colorName to color,
        outlineColorName to outlineColor,
        outlineWidthName to outlineWidthDp,
    )

    companion object {
        fun fromMap(rawMap: Any): NCircleOverlay = rawMap.asMap().let {
            NCircleOverlay(
                info = NOverlayInfo.fromMap(it[infoName]!!),
                center = it[centerName]!!.asLatLng(),
                radius = it[radiusName]!!.asDouble(),
                color = it[colorName]!!.asInt(),
                outlineColor = it[outlineColorName]!!.asInt(),
                outlineWidthDp = it[outlineWidthName]!!.asDouble(),
            )
        }

        fun fromCircleOverlay(
            circleOverlay: Overlay,
            id: String,
        ): NCircleOverlay = (circleOverlay as CircleOverlay).run {
            NCircleOverlay(
                info = NOverlayInfo(NOverlayType.CIRCLE_OVERLAY, id),
                center = center,
                radius = radius,
                color = color,
                outlineColor = outlineColor,
                outlineWidthDp = DisplayUtil.pxToDp(outlineWidth),
            )
        }


        /*
            --- Messaging Name Define ---
        */

        private const val infoName = "info"
        const val centerName = "center"
        const val radiusName = "radius"
        const val colorName = "color"
        const val outlineColorName = "outlineColor"
        const val outlineWidthName = "outlineWidth"
        const val boundsName = "bounds"
    }
}

File: ios/Classes/model/map/overlay/overlay/NCircleOverlay.swift
import NMapsMap

internal struct NCircleOverlay: AddableOverlay {
    typealias OverlayType = NMFCircleOverlay
    var overlayPayload: Dictionary<String, Any?> = [:]

    let info: NOverlayInfo
    let center: NMGLatLng
    let minZoom: Double?
    let maxZoom: Double?
    let radius: Double
    let color: UIColor
    let outlineColor: UIColor
    let outlineWidth: Double

    func createMapOverlay() -> OverlayType {
        let overlay = NMFCircleOverlay(center, radius: radius)
        if let minZoom = minZoom {
            overlay.minZoom = minZoom
        }
        if let maxZoom = maxZoom {
            overlay.maxZoom = maxZoom
        }
        overlay.fillColor = color
        overlay.outlineColor = outlineColor
        overlay.outlineWidth = outlineWidth
        return overlay
    }

    static func fromMessageable(_ v: Any) -> NCircleOverlay {
        let d = asDict(v)
        return NCircleOverlay(
                info: NOverlayInfo.fromMessageable(d[infoName]!),
                center: asLatLng(d[centerName]!),
                minZoom: castOrNull(d[minZoomName], caster: asDouble),
                maxZoom: castOrNull(d[maxZoomName], caster: asDouble),
                radius: asDouble(d[radiusName]!),
                color: asUIColor(d[colorName]!),
                outlineColor: asUIColor(d[outlineColorName]!),
                outlineWidth: asDouble(d[outlineWidthName]!)
        )
    }

    /*
        --- Messaging Name Define ---
    */
    private static let infoName = "info"
    static let centerName = "center"
    static let radiusName = "radius"
    static let colorName = "color"
    static let outlineColorName = "outlineColor"
    static let outlineWidthName = "outlineWidth"
    static let boundsName = "bounds"
}
